packages/lists/src/components/Viewport/Viewport.tsx

```diff
@@ -0,0 +1,157 @@
+  scrollDirection: Vector2D<ScrollDirection>;
+}
+
+const SCROLL_DISTANCE_ORIGIN: Vector2D<Axis> = [0, 0];
+const NO_SCROLL_DIRECTION: Vector2D<ScrollDirection> = [ScrollDirection.none, ScrollDirection.none];
+
+const SCROLLING_RESET_TIMEOUT_IN_MILLISECONDS = 200;
+
+/**
+ * Determines the scroll direction based on the given current and previous scroll distance.
+ * @param scrollDistance The new scroll distance
+ * @param prevScrollDistance The previous scroll distance
+ */
+function getScrollDirection(scrollDistance: number, prevScrollDistance: number): ScrollDirection {
+  let scrollDirection = ScrollDirection.none;
+  if (scrollDistance > prevScrollDistance) {
+    scrollDirection = ScrollDirection.forward;
+  } else if (scrollDistance < prevScrollDistance) {
+    scrollDirection = ScrollDirection.backward;
+  }
+
+  return scrollDirection;
+}
+
+/**
+ * Viewport represents a scrollable container that maintains information about its current scroll state.
+ * The component takes a function component as its child component, using the current scroll state as its only argument.
+ */
+export const Viewport = React.memo(
+  (props: IViewportProps): JSX.Element => {
+    const { height, width, children } = props;
+
+    const scrollContainerRef = useRef<HTMLDivElement>(null);
+
+    const [viewportState, setViewportState] = useState<IViewportState>({
+      isScrolling: false,
+      scrollDistance: SCROLL_DISTANCE_ORIGIN,
+      scrollDirection: NO_SCROLL_DIRECTION
+    });
+    const prevViewportState = usePreviousValue(viewportState);
+
+    const [scheduleResetScrollingTimeout, clearScrollingResetTimeout] = useDebouncedTimeout(() => {
+      setViewportState((currentViewportState: IViewportState) => {
+        return {
+          isScrolling: false,
+          scrollDistance: currentViewportState.scrollDistance,
+          scrollDirection: NO_SCROLL_DIRECTION
+        };
+      });
+    }, SCROLLING_RESET_TIMEOUT_IN_MILLISECONDS);
+
+    useEffect(() => {
+      if (scrollContainerRef.current) {
+        const onScroll = (event: Event) => {
+          // tslint:disable-next-line:no-any
+          const { scrollLeft: scrollX, scrollTop: scrollY } = ((event as any) as React.UIEvent<HTMLDivElement>).currentTarget;
+
+          const scrollDirectionX = getScrollDirection(
+            scrollX,
+            (prevViewportState && prevViewportState.scrollDistance[Axis.X]) || SCROLL_DISTANCE_ORIGIN[Axis.X]
+          );
+          const scrollDirectionY = getScrollDirection(
+            scrollY,
+            (prevViewportState && prevViewportState.scrollDistance[Axis.Y]) || SCROLL_DISTANCE_ORIGIN[Axis.Y]
+          );
+
+          setViewportState({
+            isScrolling: true,
+            scrollDistance: [scrollX, scrollY],
+            scrollDirection: [scrollDirectionX, scrollDirectionY]
+          });
+
+          scheduleResetScrollingTimeout();
+        };
+
+        scrollContainerRef.current.addEventListener<'scroll'>('scroll', onScroll);
```

Comment:
Can we mark the 'scroll' event as "passive"?

https://developers.google.com/web/updates/2016/06/passive-event-listeners

Reply:
That's a good idea, thanks!